SARP - Milestone 1 Research

masapr opened the pull request:
# Milestone Delivery Checklist

- [x] The [milestone-delivery-template.md](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md) has been copied and updated.
- [x] The [invoice form :pencil:](https://forms.gle/LSRr7PCjBpEbKGh89) has been filled out for this milestone.
- [x] This pull request is being made by the same account as the accepted application.
- [x] I have disclosed any and all sources of reused code in the submitted repositories and have done my due diligence to meet its license requirements.
- [x] In case of acceptance, the payment will be transferred to the BTC/ETH/fiat account provided in the application.
- [x] The delivery is according to the [Guidelines for Milestone Deliverables](https://grants.web3.foundation/docs/Support%20Docs/milestone-deliverables-guidelines).

Link to the application pull request: https://github.com/w3f/Grants-Program/pull/1706

niklasp commented:
Thank you @masapr for the submission of your research milestone #1. I have started an external evaluation here: #885 

Based on my evaluation, you can improve the submitted documentation and engagement (see below). I would like to see more progress in solving the major issues with MIRAI as stated in your [Findings](https://github.com/scs/MIRAI/blob/db8104d6bba5224ad445502ec46166288349e60c/substrate-examples/README.md#findings) and Open Issues [1](https://github.com/scs/MIRAI/blob/db8104d6bba5224ad445502ec46166288349e60c/substrate-examples/pallet_template/README.md#open-issues), [2](https://github.com/scs/MIRAI/blob/db8104d6bba5224ad445502ec46166288349e60c/substrate-examples/offchain-worker/README.md#open-issues) before accepting the deliverable. As I understand your granted application, the overall idea was to find if MIRAI is a suitable tool for statically analysing substrate code. By providing more details (with code examples and error outputs) it would be easier to evualuate your findings.

I am rejecting the delivery in its current state but will accept the delivery when the points stated in the evaluation are resolved or commented.

See the full evaluation in the PR above. If you have any questions or comments ask away.

masapr commented:
@niklasp thanks a lot for your evaluation.

We pushed an updated version on the [Milestone_Research1 branch](https://github.com/scs/MIRAI/blob/Milestone1_Research/substrate-examples/README.md). The main points we worked on are:
- stating clearer the [feasibility](https://github.com/scs/MIRAI/blob/074ed1dd76cfe9c371abfb9b4741c5daa5681bb2/substrate-examples/README.md#feasibility-of-approach) of using MIRAI for analysing pallet code
- adding information on how to reproduce the open issues [1](https://github.com/scs/MIRAI/blob/074ed1dd76cfe9c371abfb9b4741c5daa5681bb2/substrate-examples/pallet_template/README.md#open-issues) and [2](https://github.com/scs/MIRAI/blob/074ed1dd76cfe9c371abfb9b4741c5daa5681bb2/substrate-examples/offchain-worker/README.md#open-issues)

Apart from that, see my review comments on your [PR](https://github.com/w3f/Grant-Milestone-Delivery/pull/885#pullrequestreview-1493410427) for further details

semuelle reviewed (changes requested):
Thanks for the submission and thanks for updates, @masapr!

I see that you already addressed most of the issues raised by @niklasp. Therefore, I only have one last request: your `Documentation` deliverable is specified as, among other things, "describing the approach we plan to implement in milestone 2, incl. its limitations." I see that you added [a summary](https://github.com/scs/MIRAI/blob/074ed1dd76cfe9c371abfb9b4741c5daa5681bb2/substrate-examples/README.md#feasibility-of-approach), but what I don't understand is what the implications of your findings are. Is the work you are suggesting doable in a reasonable amount of time? Or does it justify looking into alternative implementations? And are you planning to continue this analysis? It sounds as if your [plan for a follow-up grant](https://github.com/w3f/Grants-Program/blob/master/applications/sarp-basic-functionality.md#future-plans) is to implement this further, so I assume this is something that needs to be dealt with beforehand.

masapr commented:
Thanks for your feedback @semuelle 

After the evaluation I had removed the [Next Steps](https://github.com/scs/MIRAI/blob/Milestone1_Research/substrate-examples/README.md#next-steps), as it is similar to the [Feasibility of Approach](https://github.com/scs/MIRAI/blob/Milestone1_Research/substrate-examples/README.md#feasibility-of-approach). I added it back now.

To answer your question: yes, we want to continue, with the steps I described. Overall our impression is, that MIRAI works well with "not-so-complex" code. So, if we manage to separate the substrate macro code from the pallet logic, so that MIRAI only runs on the newly created pallet logic, I'm positive that this should lead to good results. The biggest risk I see, is that we might not find a solution that Parity agrees with resp. that all solutions have too big of an impact on the substrate code base.

Apart from that, my idea was to make a cost estimate and formulate a complete proposal on this once we've got the current milestone approved. "Reasonable amount of time" is a vague term, but yes, I believe it's worth it to do this additional step and I believe it can be done in a reasonable amount of time. 

Let me know, what you think. If you're positive that we get this milestone approved and that there is a good chance we can do a follow-up grant, I'm also happy to specify the next package before getting the approval.

semuelle reviewed (approved):
Thanks for the update, @masapr. Your milestone is hereby accepted.

You can find the evaluation notes [here](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/sarp_1_niklasp.md). I have forwarded your invoice for processing.

fededubbi commented:
Hi @masapr, 

I noticed you included on the invoice both USD bank account and USDC.
In which currency do you prefer to be paid in?

Many thanks,
Federica

masapr commented:
Hi @fededubbi 

Both would be fine. But USD on the bank account would be more convenient for us.

Thank you!
Sabine

fededubbi commented:
Hi @masapr,

I noticed that you put UBS SWIFT code also with USD account, is it correct or is it another bank?

Many thanks,
Federica

masapr commented:
Hi @fededubbi 
Just to check: we haven't received anything yet. So, I'm assuming it's still to come?

fededubbi commented:
Hi @masapr,

The invoice will be paid tomorrow, but since is in USD it will take 2-3 business days for you to see it.

Many thanks,
Federica
